mantidproject/mantid
# Implement alternative space group settings

This fixes #14204.

**Changes**
Introduced `Geometry::GroupTransformation` to transform between space groups. They are specified in the same way as `SymmetryOperation` (using `x,y,z`-notation). Because of the similarities, the matrix/vector-representation was moved to a more abstract `MatrixVectorPair`-class that is used by both `SymmetryOperation` and `GroupTransformation`. The parser for the `x,y,z`-string had to be replaced because the transformation symbols are a bit different, the whole thing is now implemented using `boost::spirit`. Finally, the `SpaceGroupFactory` was extended a bit to make subscribing transformed space groups more convenient. The system test has been adjusted to test the added space groups, the documentation has been updated.

I decided not to implement a factory for the transformations. There is not much repetition happening except for subscribing the monoclinic space groups, but there are not so many. It may something to keep in mind though if people would like to transform crystal structures. In that case the factory would be useful.

**To test**
The changes regarding `MatrixVectorPair` and the parser should be covered by the existing (and new) unit tests, the other changes are covered by system- and doctests. @martyngigg, I'd appreciate if you could take a look at the parser code now that the builder code has been moved to a more appropriate place.

For manual testing you could use the following script as a starting point:

```
from mantid.geometry import SpaceGroupFactory, CrystalStructure, ReflectionGenerator

def get_hkl_d_f(crystalStructure):
    generator = ReflectionGenerator(crystalStructure)
    hkls = generator.getUniqueHKLs(0.7, 10.0)

    # Return hkl, d, F^2 sorted by (d, F^2), descending
    return sorted(zip(hkls, generator.getDValues(hkls), generator.getFsSquared(hkls)), key=lambda x: (-x[1], -x[2]))

# Two descriptions of the same crystal structure (almost silicon) with origin choice 1 and 2
almostSiliconChoice1 = CrystalStructure('5.41 5.41 5.41', 'F d -3 m', 'Si 0 0 0 1.0 0.02')
almostSiliconChoice2 = CrystalStructure('5.41 5.41 5.41', 'F d -3 m :2', 'Si 1/8 1/8 1/8 1.0 0.02')

data_1 = get_hkl_d_f(almostSiliconChoice1)
data_2 = get_hkl_d_f(almostSiliconChoice2)

# Print HKL, d and F^2
for d1, d2 in zip(data_1, data_2):
    print '{0:>10} {1:.5f} {2:8.2f} {3:>10} {4:.5f} {5:8.2f}'.format(*(d1 + d2))
```

It shows that there are now two alternative settings for No. 227 and that it does not matter which representation is chosen to do computations.

Another "pair" you can try with the above code is the following made-up structure:

```
fakeStructure1 = CrystalStructure('6 10 5', 'P b n m', 'Al 0.234 0.112 0.654 1.0 0.01; O 0 1/4 1/2 1.0 0.02')
fakeStructure2 = CrystalStructure('10 5 6', 'P n m a', 'Al 0.112 0.654 0.234 1.0 0.01; O 1/4 1/2 0 1.0 0.02')
```
